File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/ValidationUtils/ListValidation.java
package com.digicoffer.lauditor.CommonFiles.ValidationUtils;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

public class ListValidation implements TextWatcher {

    private final EditText editText;

    public ListValidation(EditText editText) {
        this.editText = editText;
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {}

    @Override
    public void afterTextChanged(Editable editable) {
        String input = editable.toString();

        if (input.isEmpty()) return;

        StringBuilder clean = new StringBuilder();

        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);

            // Reject backslash, {, }
            if (c == '\\' || c == '{' || c == '}') continue;

            if (i == 0) {
                // ❌ No space at first index
                if (c == ' ') continue;

                // ✔️ Allow digits for numbered lists
                if (Character.isDigit(c)) {
                    clean.append(c);
                    continue;
                }

                // ✔️ Allow bullet symbols (-, *, •, etc.)
                if ("•*-‣◦".indexOf(c) >= 0) {
                    clean.append(c);
                    continue;
                }

                // ✔️ Allow letters
                if (Character.isLetter(c)) {
                    clean.append(c);
                    continue;
                }

                // ✔️ Allow punctuation at start
                if (String.valueOf(c).matches("[!@#$%^&*()_+\\-=\\[\\]';:\"|,.<>?/~]")) {
                    clean.append(c);
                    continue;
                }

                // Otherwise ignore
                continue;
            }
            else {
                // Rest of characters – same as ContentValidation's "else" block
                if (String.valueOf(c).matches("[a-zA-Z0-9 !@#$%^&*()_+\\-=\\[\\]{};':\"|,.<>?/~]")) {
                    clean.append(c);
                }
            }
        }

        String finalInput = clean.toString();

        if (!finalInput.equals(input)) {
            editText.removeTextChangedListener(this);
            editText.setText(finalInput);
            editText.setSelection(finalInput.length());
            editText.addTextChangedListener(this);
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/ValidationUtils/ListValidation.kt
package com.digicoffer.lauditor.CommonFiles.ValidationUtils

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText

class ListValidation(private val editText: EditText?) : TextWatcher {

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

    override fun afterTextChanged(editable: Editable?) {
        if (editable == null) return
        val input = editable.toString()

        if (input.isEmpty()) return

        val clean = StringBuilder()

        for (i in 0 until input.length) {
            val c = input[i]

            // Reject backslash, {, }
            if (c == '\\' || c == '{' || c == '}') continue

            if (i == 0) {
                // No space at first index
                if (c == ' ') continue

                // Allow digits for numbered lists
                if (Character.isDigit(c)) {
                    clean.append(c)
                    continue
                }

                // Allow bullet symbols (-, *, •, etc.)
                if ("•*-‣◦".indexOf(c) >= 0) {
                    clean.append(c)
                    continue
                }

                // Allow letters
                if (Character.isLetter(c)) {
                    clean.append(c)
                    continue
                }

                // Allow punctuation at start
                if (c.toString().matches("[!@#$%^&*()_+\\-=\\[\\]';:\"|,.<>?/~]".toRegex())) {
                    clean.append(c)
                    continue
                }

                // Otherwise ignore
                continue
            } else {
                // Rest of characters – same as ContentValidation's "else" block
                if (c.toString().matches("[a-zA-Z0-9 !@#$%^&*()_+\\-=\\[\\]{};':\"|,.<>?/~]".toRegex())) {
                    clean.append(c)
                }
            }
        }

        val finalInput = clean.toString()

        if (finalInput != input) {
            editText?.removeTextChangedListener(this)
            editText?.setText(finalInput)
            editText?.setSelection(finalInput.length)
            editText?.addTextChangedListener(this)
        }
    }
}
